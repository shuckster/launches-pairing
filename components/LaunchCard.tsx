import type { TLaunchCardSchema } from '../schemas/ui/launch-card.types';

import Image from 'next/image';
import { Card } from './Card';
import { LaunchStatus } from './LaunchStatus';
import { LaunchDate } from './LaunchDate';
import { TextLabel } from './TextLabel';

type TLaunchCoreSerialProps = { coreSerial: string };
function LaunchCoreSerial(props: TLaunchCoreSerialProps) {
  return <TextLabel label="Core serial">{props.coreSerial}</TextLabel>;
}

type TLaunchPayloadTypeProps = { payloadType: string };
function LaunchPayloadType(props: TLaunchPayloadTypeProps) {
  return <TextLabel label="Payload type">{props.payloadType}</TextLabel>;
}

type TLaunchCardProps = { data: TLaunchCardSchema };
export function LaunchCard(props: TLaunchCardProps) {
  const { data } = props;

  if (!data.name) {
    /**
     * LaunchCard: no name for card; is the rest of the data there?
     * @todo - log to Sentry, throw error to be captured by ErrorBoundary, etc.
     */
    return null;
  }

  return (
    <Card
      className="launch-card"
      ImageSlot={
        data.patchUrl && (
          <Image
            src={data.patchUrl}
            alt={`Insignia for ${data.name}`}
            width={125}
            height={125}
          />
        )
      }
      HeadingSlot={<>{data.name}</>}
      BodySlot={
        <>
          {data.coreSerial && <LaunchCoreSerial coreSerial={data.coreSerial} />}
          {data.payloadType && (
            <LaunchPayloadType payloadType={data.payloadType} />
          )}
          {
            <LaunchStatus
              success={data.success}
              failureReason={data.failureReason}
            />
          }
          {data.dateUtc && (
            <LaunchDate dateUtc={data.dateUtc} success={data.success} />
          )}
        </>
      }
    />
  );
}
