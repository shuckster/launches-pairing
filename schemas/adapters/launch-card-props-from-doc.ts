import type { TPartialLaunchesDocumentSchema } from '../api/spacex.types';
import type { TLaunchCardSchema } from '../ui/launch-card.types';

/**
 * input is expected to be TLaunchesDocumentSchema, but since
 * it comes from an external API, we can't guarantee that.
 *
 * Could also use Zod et al for schema validation, but trying
 * to keep it basic for this task.
 */

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function launchCardPropsFromApiDoc(apiDoc: any): TLaunchCardSchema {
  const input = (apiDoc || {}) as TPartialLaunchesDocumentSchema;
  return {
    id: input.id ?? null,
    name: input.name ?? null,
    dateUtc: input.date_utc ?? null,
    coreSerial: input.cores?.[0]?.core?.serial ?? null,
    payloadType: input.payloads?.[0]?.type ?? null,
    patchUrl: input.links?.patch?.small ?? null,
    success: input.success ?? null,
    failureReason: input.failures?.[0]?.reason ?? null,
  };
}
