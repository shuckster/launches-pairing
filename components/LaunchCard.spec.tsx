import { render, screen } from '@testing-library/react';
import { LaunchCard } from './LaunchCard';

import mockData from '@/pages/api/launches/mock.json';
import { launchCardPropsFromApiDoc } from '@/schemas/adapters/launch-card-props-from-doc';

const cards = (mockData.docs ?? []).map(launchCardPropsFromApiDoc);
const successCard = cards.find((card) => card.success === true);
const failureCard = cards.find((card) => card.success === false);

/**
 * Happy paths: renders card with complete data
 */

describe('<LaunchCard />', () => {
  it('renders nothing if we do not have at least a name', () => {
    const { asFragment } = render(
      <LaunchCard data={{ id: '1', name: null, success: null }} />
    );

    const heading = screen.queryByRole('heading', { name: successCard.name });
    expect(heading).not.toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders a card with successful launch data', () => {
    const { asFragment } = render(<LaunchCard data={successCard} />);

    const heading = screen.getByRole('heading', { name: successCard.name });
    const serial = screen.getByText(successCard.coreSerial);
    const payload = screen.getByText(successCard.payloadType);
    const status = screen.getByText('Success');

    expect(heading).toBeInTheDocument();
    expect(serial).toBeInTheDocument();
    expect(payload).toBeInTheDocument();
    expect(status).toBeInTheDocument();

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders a card with failed launch data', () => {
    const { asFragment } = render(<LaunchCard data={failureCard} />);

    const heading = screen.getByRole('heading', { name: failureCard.name });
    const serial = screen.getByText(failureCard.coreSerial);
    const payload = screen.getByText(failureCard.payloadType);
    const status = screen.getByText('Failure:');

    expect(heading).toBeInTheDocument();
    expect(serial).toBeInTheDocument();
    expect(payload).toBeInTheDocument();
    expect(status).toBeInTheDocument();

    expect(asFragment()).toMatchSnapshot();
  });
});
