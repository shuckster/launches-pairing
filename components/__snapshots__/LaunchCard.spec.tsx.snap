// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LaunchCard /> renders a card with failed launch data 1`] = `
<DocumentFragment>
  <figure
    class="launch-card day p-6 max-w-sm md:m-auto bg-slate-50 dark:bg-slate-600 rounded-lg text-black dark:text-white shadow-lg flex flex-col space-y-4 md:flex-row md:space-x-4"
  >
    <div
      class="shrink-0 self-center"
    >
      <img
        alt="Insignia for FalconSat"
        data-nimg="1"
        decoding="async"
        height="125"
        loading="lazy"
        src="/_next/image?url=https%3A%2F%2Fimages2.imgbox.com%2F94%2Ff2%2FNN6Ph45r_o.png&w=256&q=75"
        srcset="/_next/image?url=https%3A%2F%2Fimages2.imgbox.com%2F94%2Ff2%2FNN6Ph45r_o.png&w=128&q=75 1x, /_next/image?url=https%3A%2F%2Fimages2.imgbox.com%2F94%2Ff2%2FNN6Ph45r_o.png&w=256&q=75 2x"
        style="color: transparent;"
        width="125"
      />
    </div>
    <div
      class="text-gray-500 dark:text-gray-400"
    >
      <h3
        class="text-gray-700 dark:text-gray-100 text-xl font-semibold"
      >
        FalconSat
      </h3>
      <dl
        class="mt-1 mb-2"
      >
        <dt
          class="text-xs"
        >
          Core serial:
        </dt>
        <dd
          class="text-gray-700 dark:text-gray-200"
        >
          Merlin1A
        </dd>
      </dl>
      <dl
        class="mt-1 mb-2"
      >
        <dt
          class="text-xs"
        >
          Payload type:
        </dt>
        <dd
          class="text-gray-700 dark:text-gray-200"
        >
          Satellite
        </dd>
      </dl>
      <div
        class="rounded-t-lg bg-red-100 dark:bg-opacity-20 dark:bg-red-900 dark:text-orange-200 py-2 px-3 mt-2 inline-block text-sm"
      >
        <strong
          class="dark:text-white"
        >
          Failure:
        </strong>
         merlin engine failure
      </div>
      <div
        class="rounded-b-lg bg-gray-100 dark:bg-gray-700 text-gray-500 dark:text-white px-3 py-2 border-t-2 text-sm border-red-500"
      >
        <span
          class="whitespace-nowrap"
        >
          24 March 2006
        </span>
         
        <span
          class="whitespace-nowrap"
        >
          - 22:30:00 GMT
        </span>
      </div>
    </div>
  </figure>
</DocumentFragment>
`;

exports[`<LaunchCard /> renders a card with successful launch data 1`] = `
<DocumentFragment>
  <figure
    class="launch-card day p-6 max-w-sm md:m-auto bg-slate-50 dark:bg-slate-600 rounded-lg text-black dark:text-white shadow-lg flex flex-col space-y-4 md:flex-row md:space-x-4"
  >
    <div
      class="shrink-0 self-center"
    >
      <img
        alt="Insignia for RatSat"
        data-nimg="1"
        decoding="async"
        height="125"
        loading="lazy"
        src="/_next/image?url=https%3A%2F%2Fimages2.imgbox.com%2F95%2F39%2FsRqN7rsv_o.png&w=256&q=75"
        srcset="/_next/image?url=https%3A%2F%2Fimages2.imgbox.com%2F95%2F39%2FsRqN7rsv_o.png&w=128&q=75 1x, /_next/image?url=https%3A%2F%2Fimages2.imgbox.com%2F95%2F39%2FsRqN7rsv_o.png&w=256&q=75 2x"
        style="color: transparent;"
        width="125"
      />
    </div>
    <div
      class="text-gray-500 dark:text-gray-400"
    >
      <h3
        class="text-gray-700 dark:text-gray-100 text-xl font-semibold"
      >
        RatSat
      </h3>
      <dl
        class="mt-1 mb-2"
      >
        <dt
          class="text-xs"
        >
          Core serial:
        </dt>
        <dd
          class="text-gray-700 dark:text-gray-200"
        >
          Merlin2C
        </dd>
      </dl>
      <dl
        class="mt-1 mb-2"
      >
        <dt
          class="text-xs"
        >
          Payload type:
        </dt>
        <dd
          class="text-gray-700 dark:text-gray-200"
        >
          Satellite
        </dd>
      </dl>
      <div
        class="rounded-t-lg text-lime-700 bg-lime-200 dark:bg-lime-900 dark:text-lime-100 py-1 px-3 mt-2 inline-block font-bold"
      >
        Success
      </div>
      <div
        class="rounded-b-lg bg-gray-100 dark:bg-gray-700 text-gray-500 dark:text-white px-3 py-2 border-t-2 text-sm border-lime-700"
      >
        <span
          class="whitespace-nowrap"
        >
          29 September 2008
        </span>
         
        <span
          class="whitespace-nowrap"
        >
          - 00:15:00 BST
        </span>
      </div>
    </div>
  </figure>
</DocumentFragment>
`;

exports[`<LaunchCard /> renders nothing if we do not have at least a name 1`] = `<DocumentFragment />`;
